File: src/unity/Runtime/Core/ObserverHandle.cs
using System;
using System.Collections.Generic;

namespace EE {
    public class ObserverHandle {
        private readonly List<ScopeGuard> _guards;

        public ObserverHandle() {
            _guards = new List<ScopeGuard>();
        }

        public ObserverBinder<Observer> Bind<Observer>(IObserverManager<Observer> manager) {
            return new ObserverBinder<Observer>(this, manager);
        }

        internal void Add(Action rollback) {
            _guards.Add(new ScopeGuard(rollback));
        }

        public void Clear() {
            foreach (var entry in _guards) {
                entry.Dispose();
            }
            _guards.Clear();
        }
    }
}

File: src/ts/src/core/ObserverHandle.ts
import { IObserverManager } from "./IObserverManager";
import { ObserverBinder } from "./ObserverBinder";
import { ScopeGuard } from "./ScopeGuard";

export class ObserverHandle {
    private readonly _guards: ScopeGuard[] = [];

    public bind<Observer>(manager: IObserverManager<Observer>): ObserverBinder<Observer> {
        return new ObserverBinder<Observer>(this, manager);
    }

    public add(rollback: () => void) {
        this._guards.push(new ScopeGuard(rollback));
    }

    public clear(): void {
        for (const entry of this._guards) {
            entry.dispose();
        }
        this._guards.length = 0;
    }
}